### PR TITLE
agent: Remove `enable-remote-node-identity` flag

### PR DESCRIPTION
**Pull request by doniacld**
Remove the flag `enable-remote-node-identity` after being deprecated in v1.15. 
Update the daemon, manager and associated tests.

```release-note
Remove helm option `enable-remote-node-identity` after being deprecated in v1.15.
```

### CODE REVIEW

File: daemon/cmd/daemon_main.go
Comment (gandro) [inline daemon/cmd/daemon_main.go:285]: Please update the `cmdref` docs (by running `make -C Documentation update-cmdref`) in the same commit which removes the flag. Thanks!

File: contrib/vagrant/start.sh
Comment (gandro) [inline contrib/vagrant/start.sh:310]: Consider moving this commit to the beginning of the git history. Using a flag which is removed a previous commit might break `git bisect`.

File: install/kubernetes/cilium/templates/cilium-configmap.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml:854]: Same as above: Consider moving this commit to the beginning of the git history. Using a flag which is removed a previous commit might break git bisect.

File: Documentation/cmdref/cilium-agent.md
Comment (gandro) [inline Documentation/cmdref/cilium-agent.md]: Something is wrong here. This commit removes all `cmdref` documentation, instead of regenerating it. I'd recommend dropping this commit and doing the `cmdref` change in the commit which also removes the flag in `daemon_main.go`.

File: pkg/node/manager/manager.go
Comment (gandro) [inline pkg/node/manager/manager.go:406]: This looks incomplete. `m.conf.RemoteNodeIdentitiesEnabled()` reads (at this point) non-existant value. Consider moving this commit to the earlier in the git history and make sure to also remove the `RemoteNodeIdentitiesEnabled()` function itself.

File: pkg/node/manager/manager.go
Comment (gandro) [inline pkg/node/manager/manager.go:399]: This file also reads `RemoteNodeIdentitiesEnabled` a bit below in `nodeIdentityLabels`. That usage also needs to be removed.

File: pkg/node/manager/manager.go
Comment (gandro) [inline pkg/node/manager/manager.go:407]: This looks incorrect. `m.conf.RemoteNodeIdentitiesEnabled()` is true by default. This means that this function here basically always returned false, and should now statically return false. However, as is, it defaults to returning true.

Let's remove fully remove `legacyNodeIpBehavior`, as we can now assume that is always returns false.

**Review by gandro — CHANGES_REQUESTED** [review]
Thanks! Overall this is a good cleanup, but there are some issues with the git history. In particular, we want each commit itself to build and ideally also pass CI. In order to achieve this, I think you need to reorder some of the commits.

**Comment by gandro** [comment]
I also wonder if we should call this out in the upgrade docs. I do hope we no longer have users running with `enable-remote-node-identity: false`, but if they do they will have to expect drops during upgrade.

File: pkg/node/manager/manager.go
Comment (doniacld) [reply pkg/node/manager/manager.go:407]: Done

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (gandro) [inline install/kubernetes/cilium/values.yaml.tmpl:2219]: I think this might needs the docs to be regenerated with `make -C Documentation update-helm-values`

File: pkg/node/manager/manager.go
Comment (gandro) [inline pkg/node/manager/manager.go]: This function is now unused, correct? Let's remove it.

File: pkg/option/config.go
Comment (gandro) [inline pkg/option/config.go]: Any reason we can't remove this field here?

**Review by gandro — CHANGES_REQUESTED** [review]
Thanks, looking much better already. Still a few minor things that should be addressed

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (doniacld) [reply install/kubernetes/cilium/values.yaml.tmpl:2219]: Done

File: Documentation/gettingstarted/terminology.rst
Comment (gandro) [inline Documentation/gettingstarted/terminology.rst]: It's enabled by default since Cilium v1.7.

I think we should remove the whole paragraph however, it doesn't make sense anymore really.

File: pkg/ipcache/kvstore.go
Comment (gandro) [inline pkg/ipcache/kvstore.go]:
```suggestion
```

Comment (gandro) [reply pkg/ipcache/kvstore.go]: Though I wonder if we should just remove the if statement. After all, we don't expect remote nodes to announce themselves as "host" anymore.

File: pkg/policy/mapstate.go
Comment (gandro) [inline pkg/policy/mapstate.go:1309]: `var localRemoteNodeKey` is unused now and can be removed

File: pkg/ipcache/kvstore.go
Comment (doniacld) [reply pkg/ipcache/kvstore.go]: True!

File: Documentation/gettingstarted/terminology.rst
Comment (gandro) [reply Documentation/gettingstarted/terminology.rst]: This was marked as resolve but the paragraph is still here?

File: pkg/ipcache/kvstore.go
Comment (nathanjsweet) [inline pkg/ipcache/kvstore.go:315]: This should still compare `peerIdentity` to `identity.ReservedIdentityHost` and set `peerIdentity` to `ReservedIdentityRemoteNode` if they are equal.
```suggestion
if peerIdentity == identity.ReservedIdentityHost {
```

**Review by nathanjsweet — CHANGES_REQUESTED** [review]

File: pkg/ipcache/kvstore.go
Comment (nathanjsweet) [reply pkg/ipcache/kvstore.go]: >Though I wonder if we should just remove the if statement. After all, we don't expect remote nodes to announce themselves as "host" anymore.

@gandro Where is this logic? I can't find it. I can only find that [`UpsertIPToKVStore` appears to upsert host IPs with unmodified IDs](https://github.com/cilium/cilium/blob/687e4f8f403aa75f3e1e42005db51b29fdd9e5ef/pkg/endpoint/policy.go#L836-L850). Am I missing something?

Comment (gandro) [reply pkg/ipcache/kvstore.go]: @nathanjsweet Yeah what I said was wrong. I looked at the [old migration docs](https://docs.cilium.io/en/v1.7/install/upgrade/#configmap-remote-node-identity) and they mentioned (the long removed) `transmit-host-src`, which I thought meant impacted IPCache entries, but it doesn't. It only affected the identities transmitted in VXLAN headers.

However, I'm not sure what host IPs actually announce in the linked endpoint sync controller. 

I assume external workloads will announce regular identity that is different for `reserved:host` (i.e. the one read by `e.SecurityIdentity.ID` allocated when the EW joins the cluster). And for the host endpoint I believe we don't run the `runIPIdentitySync` controller either?

In any case, thinking about this, I'm happy to keep the if statement there, it's a good safeguard against bugs in the remote agents in case they ever attempt to create an kvstore entry with host identity.

**Review by nathanjsweet — APPROVED** [review]
Thanks!

File: Documentation/operations/upgrade.rst
Comment (gandro) [inline Documentation/operations/upgrade.rst]:
```suggestion
* The deprecated helm option ``remoteNodeIdentity`` has been removed. This should have no impact which used the previous default value of ``true``: Remote nodes will now always use `remote-node` identity.
  If you have network policies based on ``enable-remote-node-identity=false`` make sure to update them.
```

File: install/kubernetes/cilium/README.md
Comment (gandro) [inline install/kubernetes/cilium/README.md]: Something is off here, `resourceQuotas` is supposed to be on a new line.

File: daemon/cmd/daemon_main.go
Comment (gandro) [reply daemon/cmd/daemon_main.go:285]: This is not resolved and still needs cmdref to be regenerated.

**Review by gandro — COMMENTED** [review]
Starting to look good. A few docs nits.

File: daemon/cmd/daemon_main.go
Comment (gandro) [reply daemon/cmd/daemon_main.go:285]: Ah, it was already hidden, no need to regenerate the cmdref.

File: Documentation/operations/upgrade.rst
Comment (gandro) [inline Documentation/operations/upgrade.rst]:
```suggestion
* The deprecated Helm option ``remoteNodeIdentity`` has been removed. This should have no impact on users which used the previous default
```

File: USERS.md
Comment (gandro) [inline USERS.md]: Unrelated change

File: Documentation/operations/upgrade.rst
Comment (gandro) [inline Documentation/operations/upgrade.rst]: Docs build is failing due to `remote-node` not using double quotes:

    `remote-nodes` -> ``remote-nodes``

**Review by gandro — COMMENTED** [review]
Code-wise looks good to me now! A couple of docs nits still.

Also, grepping for `enable-remote-node-identity` I've also found those files:

https://github.com/cilium/cilium/blob/main/examples/policies/host/lock-down-dev-vms-cidr-node.yaml
https://github.com/cilium/cilium/blob/main/examples/policies/host/lock-down-dev-vms-remote-node.yaml

We should delete the first one and remove the mention of the flag from the second one.

File: Documentation/operations/upgrade.rst
Comment (derailed) [reply Documentation/operations/upgrade.rst]: `on users who used the previous default` 
A bit out of context here... from what I can tell this option defaulted to `true`. Based on this doc would there be impact then if users had unset it i.e remoteNodeIdentity=false?

**Review by derailed — APPROVED** [review]
@doniacld Nice cleanup Donia!!

File: Documentation/operations/upgrade.rst
Comment (gandro) [reply Documentation/operations/upgrade.rst]: Correct, users had to manually set it to `false` to have an upgrade impact. If no value was not provided, it defaulted to true.

**Review by gandro — APPROVED** [review]
Awesome, thank you!

File: pkg/policy/api/entity.go
Comment (tklauser) [inline pkg/policy/api/entity.go:150]: If I'm not mistaken, we should still keep `endpointSelectorHost` here, i.e.

```go
EntitySelectorMapping[EntityHost] = EndpointSelectorSlice{endpointSelectorHost}
```

File: pkg/policy/rule.go
Comment (tklauser) [inline pkg/policy/rule.go:422]: nit:

```suggestion
	hostWildcardL7 := make([]string, 0, 1)
```

as we're not going to add the second item to the slice anymore in any case.

**Review by tklauser — CHANGES_REQUESTED** [review]
Thanks for cleaning this up.

Two comments inline, one need to be addressed I think and the other is more nitpicking than anything else :)

Also, I think you need to remove the line https://github.com/cilium/cilium/blob/b58125d885edbb278f11f84303c0e7c934ca7ea4/.github/actions/ginkgo/main-focus.yaml#L66 to fix the [failing GH action validation workflow](https://github.com/cilium/cilium/actions/runs/8283348471/job/22666320317?pr=31228) in CI.

**Review by danehans — COMMENTED** [review]
Aside from the comments that need to be resolved from @tklauser, LGTM

File: pkg/policy/api/entity.go
Comment (doniacld) [reply pkg/policy/api/entity.go:150]: `EntitySelectorMapping[EntityHost]` is already filled in the global variable here https://github.com/cilium/cilium/blob/main/pkg/policy/api/entity.go#L95. 
This part of the code was only present to handle the remote node as a host usecase, no need to keep it.

File: pkg/policy/rule.go
Comment (tklauser) [reply pkg/policy/rule.go:422]: Sorry, I don't see which comment specifically you mean. This is not related to my previous comment AFAICT.

Comment (tklauser) [reply pkg/policy/rule.go:422]: In any case it's just a nitpick and not worth a respin. Resolving.

**Review by tklauser — APPROVED** [review]
Thanks Donia!

**Review by ghost — APPROVED** [review]

**Review by viktor-kurchenko — APPROVED** [review]

**Review by aditighag — APPROVED** [review]
Looks good for my CODEOWNERS

**Review by gandro — APPROVED** [review]
Re-reviewed, looks good to me! Thanks a lot

(I've reviewed the changes to `manager_test.go` in Goland, since the diff GitHub produces is borderline unreadable)

**Review by ysksuzuki — APPROVED** [review]

**Review by danehans — APPROVED** [review]
@doniacld this PR needs a rebase. Otherwise, LGTM.

**Comment by gandro** [comment]
Needs another rebase due to conflicts

**Pull request merged**
